### PR TITLE
Don't transition Task to STOPPED when StopContainer times out.

### PR DESCRIPTION
**Pull request by aaithal**
Transitioning the last_status of the Task to STOPPED and emitting the same
to ECS backend without actually stopping the container can result in a
scenario where the container is still running on an instance while the
Task is marked as STOPPED. This change offsets the Task transition to occur
when the engine gets terminal events for the container from docker event
stream instead.

Verified that unit, integ and functional tests pass. 

r? @samuelkarp @juanrhenals @richardpen

### CODE REVIEW

File: agent/engine/docker_task_engine_test.go [hunk 1]
```diff
@@ -299,8 +298,8 @@ func TestStartTimeoutThenStart(t *testing.T) {
 
 		// Expect it to try to stop the container before going on;
 		// in the future the agent might optimize to not stop unless the known
-		// status is running, at which poitn this can be safeuly removed
-		client.EXPECT().StopContainer("containerId").Return(DockerContainerMetadata{Error: errors.New("Cannot start")})
+		// status is running, at which point this can be safeuly removed
```
Comment (samuelkarp) [inline agent/engine/docker_task_engine_test.go]: s/safeuly/safely as well if you're fixing typos :)

Comment (aaithal) [inline agent/engine/docker_task_engine_test.go]: done

File: agent/engine/docker_task_engine_test.go [hunk 2]
```diff
@@ -513,6 +512,87 @@ func TestCreateContainerForceSave(t *testing.T) {
...
+			go func() { eventStream <- dockerEvent(api.ContainerCreated) }()
+		}).Return(DockerContainerMetadata{DockerId: "containerId"})
+
+		// StartContainer returns timeout error. This should cause the engine
+		// to transition the task to STOPPED and to stop all containers of
+		// the task
+		client.EXPECT().StartContainer("containerId").Return(DockerContainerMetadata{Error: &DockerTimeoutError{}})
+
+		// StopContainer times out as well
+		client.EXPECT().StopContainer(gomock.Any()).Do(func(id string) {
```
Comment (samuelkarp) [inline agent/engine/docker_task_engine_test.go]: Can you put the two `StopContainer` expectations inside a `gomock.InOrder` to enforce ordering?

File: agent/engine/errors.go [hunk 3]
```diff
@@ -97,3 +106,15 @@ func (c CannotGetDockerClientError) Error() string {
 func (CannotGetDockerClientError) ErrorName() string {
 	return "CannotGetDockerclientError"
 }
+
+type CannotAuthRegistryError struct {
```
Comment (samuelkarp) [inline agent/engine/errors.go]: Why not reuse `CannotXContainerError` with Pull?

File: agent/engine/task_manager.go [hunk 4]
```diff
@@ -227,21 +227,35 @@ func (mtask *managedTask) handleContainerChange(containerChange dockerContainerC
...
+			// This ensures that we don't emit a containerstopped event; a
+			// terminal container event from docker event stream will instead be
+			// responsible for the transition. Alternatively, the steadyState check
+			// could also trigger the progress and have another go at stopping the
+			// container
+			if event.Error.ErrorName() == dockerTimeoutErrorName {
+				seelog.Infof("%s for 'docker stop' of container; ignoring state change;  task: %v, container: %v, error: %v", dockerTimeoutErrorName, mtask.Task, container, event.Error.Error())
+				container.KnownStatus = currentKnownStatus
+				return
+			} else {
```
Comment (samuelkarp) [inline agent/engine/task_manager.go]: Can we special-case the empty-host-volume fake-container-thing so we don't try to stop it (it never runs) and don't spew errors in the log too?

File: agent/engine/errors.go (hunk 3 above)
Comment (aaithal) [inline agent/engine/errors.go]: I like that. Will modify.

File: agent/engine/task_manager.go (hunk 4 above)
Comment (aaithal) [inline agent/engine/task_manager.go]: sure, will do

File: agent/engine/docker_task_engine_test.go (hunk 2 above)
Comment (aaithal) [inline agent/engine/docker_task_engine_test.go]: done

File: agent/engine/task_manager.go (hunk 4 above)
Comment (aaithal) [inline agent/engine/task_manager.go]: On second thought, I'd like that to be addressed in a separate PR. Want to get this bug fix out of the way.

**Comment by aaithal** [comment]
@samuelkarp thanks for taking a look, I have addressed your comments.

**Comment by samuelkarp** [comment]
:+1: Let's let this run for a bit under load and see how it performs.

**Comment by samuelkarp** [comment]
@aaithal Looks like this needs a rebase.

**Comment by aaithal** [comment]
Verified that unit, integration and functional tests pass after rebasing.

**Comment by samuelkarp** [comment]
:+1:
